// tower-idle-toolkit/build/constants/bots.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BOTS = exports.FLAME_BOT = exports.THUNDER_BOT = exports.AMPLIFY_BOT = exports.GOLDEN_BOT = exports.BOT_UNLOCK_COSTS = void 0;
/**
 * @constant BOT_UNLOCK_COSTS - Medal costs to unlock each successive bot
 */
exports.BOT_UNLOCK_COSTS = [150, 300, 600, 900];
const frequencyFormula = (base) => (level, lab = 0) => (base - 0.3 * level) * (1 - lab / 100);
const valueFormula = (base = 0, growth = 0) => (level) => base + growth * level;
const formatBonus = (bonus) => `${bonus.toFixed(2)}x`;
const formatRange = (range) => `${range.toFixed(2)}m`;
const formatDuration = (duration) => `${duration.toFixed(2)}s`;
/**
 * @constant GOLDEN_BOT - Golden Bot
 */
exports.GOLDEN_BOT = Object.freeze({
    Bonus: Object.freeze({
        baseValue: 2,
        medalUpgrades: 20,
        getValue: valueFormula(2, 0.2),
        formatValue: formatBonus,
    }),
    Frequency: Object.freeze({
        baseValue: 25,
        medalUpgrades: 20,
        labUpgrades: 20,
        getValue: frequencyFormula(25),
        formatValue: formatDuration,
    }),
    Range: Object.freeze({
        baseValue: 20,
        medalUpgrades: 15,
        getValue: valueFormula(20, 2),
        formatValue: formatRange,
    }),
});
/**
 * @constant AMPLIFY_BOT - Amplify Bot
 */
exports.AMPLIFY_BOT = Object.freeze({
    Bonus: Object.freeze({
        baseValue: 3.5,
        medalUpgrades: 20,
        getValue: valueFormula(2, 0.4),
        formatValue: formatBonus,
    }),
    Frequency: Object.freeze({
        baseValue: 25,
        medalUpgrades: 20,
        labUpgrades: 20,
        getValue: frequencyFormula(25),
        formatValue: formatDuration,
    }),
    Range: Object.freeze({
        baseValue: 25,
        medalUpgrades: 15,
        getValue: valueFormula(25, 2),
        formatValue: formatRange,
    }),
});
/**
 * @constant THUNDER_BOT - Thunder Bot
 */
exports.THUNDER_BOT = Object.freeze({
    Duration: Object.freeze({
        baseValue: 4,
        medalUpgrades: 20,
        getValue: valueFormula(4, 0.4),
        formatValue: formatBonus,
    }),
    Frequency: Object.freeze({
        baseValue: 13,
        medalUpgrades: 15,
        labUpgrades: 20,
        getValue: frequencyFormula(13),
        formatValue: formatDuration,
    }),
    Range: Object.freeze({
        baseValue: 25,
        medalUpgrades: 15,
        getValue: valueFormula(25, 3),
        formatValue: formatRange,
    }),
});
/**
 * @constant FLAME_BOT - Flame Bot
 */
exports.FLAME_BOT = Object.freeze({
    Damage: Object.freeze({
        baseValue: 5000,
        medalUpgrades: 20,
        getValue: valueFormula(5000, 800),
        formatValue: (value) => `${value.toFixed(0)}%`,
    }),
    Frequency: Object.freeze({
        baseValue: 8,
        medalUpgrades: 15,
        labUpgrades: 20,
        getValue: frequencyFormula(8),
        formatValue: formatDuration,
    }),
    Range: Object.freeze({
        baseValue: 30,
        medalUpgrades: 15,
        getValue: valueFormula(30, 4),
        formatValue: formatRange,
    }),
});
/**
 * @constant BOTS - All of the available bots as an object mapping their In Game English Name to the config for that bot.
 */
exports.BOTS = Object.freeze({
    'Golden Bot': exports.GOLDEN_BOT,
    'Amplify Bot': exports.AMPLIFY_BOT,
    'Thunder Bot': exports.THUNDER_BOT,
    'Flame Bot': exports.FLAME_BOT,
});
