// tower-idle-toolkit/build/constants/cards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LOCKED_CARDS = exports.CARDS = exports.EPIC_CARDS = exports.ULTIMATE_CRIT_CARD = exports.BERSERKER_CARD = exports.WAVE_ACCELERATOR_CARD = exports.ENERGY_SHIELD_CARD = exports.DEMON_MODE_CARD = exports.SECOND_WIND_CARD = exports.SUPER_TOWER_CARD = exports.ENERGY_NET_CARD = exports.DEATH_RAY_CARD = exports.RARE_CARDS = exports.RECOVERY_PACKAGE_CHANCE_CARD = exports.LAND_MINE_STUN_CARD = exports.FREE_UPGRADES_CARD = exports.EXTRA_ORB_CARD = exports.PLASMA_CANNON_CARD = exports.CRIT_COIN_CARD = exports.WAVE_SKIP_CARD = exports.INTRO_SPRINT_CARD = exports.COMMON_CARDS = exports.FORTRESS_CARD = exports.EXTRA_DEFENSE_CARD = exports.ENEMY_BALANCE_CARD = exports.CRIT_CHANCE_CARD = exports.SLOW_AURA_CARD = exports.COIN_CARD = exports.CASH_CARD = exports.RANGE_CARD = exports.REGEN_CARD = exports.HEALTH_CARD = exports.ASPD_CARD = exports.DAMAGE_CARD = void 0;
/**
 * @constant DAMAGE_CARD - Maps Damage Card stars to value, 0 is unequipped
 */
exports.DAMAGE_CARD = Object.freeze({
    0: 1,
    1: 1.5,
    2: 2,
    3: 2.4,
    4: 2.8,
    5: 3.2,
    6: 3.6,
    7: 4,
});
/**
 * @constant ASPD_CARD - Maps Attack Speed Card stars to value, 0 is unequipped
 */
exports.ASPD_CARD = Object.freeze({
    0: 1,
    1: 1.25,
    2: 1.4,
    3: 1.55,
    4: 1.7,
    5: 1.85,
    6: 2,
    7: 2.15,
});
/**
 * @constant HEALTH_CARD - Maps Health Card stars to value, 0 is unequipped
 */
exports.HEALTH_CARD = Object.freeze({
    0: 1,
    1: 1.5,
    2: 2,
    3: 2.4,
    4: 2.8,
    5: 3.2,
    6: 3.6,
    7: 4,
});
/**
 * @constant REGEN_CARD - Maps Regen Card stars to value, 0 is unequipped
 */
exports.REGEN_CARD = Object.freeze({
    0: 1,
    1: 1.4,
    2: 1.6,
    3: 1.8,
    4: 2,
    5: 2.2,
    6: 2.4,
    7: 2.6,
});
/**
 * @constant RANGE_CARD - Maps Range Card stars to value, 0 is unequipped
 */
exports.RANGE_CARD = Object.freeze({
    0: 1,
    1: 1.15,
    2: 1.2,
    3: 1.25,
    4: 1.3,
    5: 1.35,
    6: 1.4,
    7: 1.45,
});
/**
 * @constant CASH_CARD - Maps Cash Card stars to value, 0 is unequipped
 */
exports.CASH_CARD = Object.freeze({
    0: 1,
    1: 1.2,
    2: 1.4,
    3: 1.6,
    4: 1.8,
    5: 2,
    6: 2.2,
    7: 2.4,
});
/**
 * @constant COIN_CARD - Maps Coin Card stars to value, 0 is unequipped
 */
exports.COIN_CARD = Object.freeze({
    0: 1,
    1: 1.15,
    2: 1.2,
    3: 1.25,
    4: 1.3,
    5: 1.35,
    6: 1.4,
    7: 1.45,
});
/**
 * @constant SLOW_AURA_CARD - Maps Slow Aura Card stars to value, 0 is unequipped
 */
exports.SLOW_AURA_CARD = Object.freeze({
    0: 0,
    1: 13,
    2: 16,
    3: 19,
    4: 22,
    5: 25,
    6: 28,
    7: 31,
});
/**
 * @constant CRIT_CHANCE_CARD - Maps Crit Chance Card stars to value, 0 is unequipped
 */
exports.CRIT_CHANCE_CARD = Object.freeze({
    0: 0,
    1: 5,
    2: 6,
    3: 7,
    4: 8,
    5: 9,
    6: 10,
    7: 11,
});
/**
 * @constant ENEMY_BALANCE_CARD - Maps Enemy Balance Card stars to value, 0 is unequipped
 */
exports.ENEMY_BALANCE_CARD = Object.freeze({
    0: 1,
    1: 1.3,
    2: 1.4,
    3: 1.5,
    4: 1.6,
    5: 1.7,
    6: 1.8,
    7: 1.9,
});
/**
 * @constant EXTRA_DEFENSE_CARD - Maps Extra Defense Card stars to value, 0 is unequipped
 */
exports.EXTRA_DEFENSE_CARD = Object.freeze({
    0: 0,
    1: 5,
    2: 6,
    3: 7,
    4: 8,
    5: 9,
    6: 10,
    7: 11,
});
/**
 * @constant FORTRESS_CARD - Maps Fortress Card stars to value, 0 is unequipped
 */
exports.FORTRESS_CARD = Object.freeze({
    0: 1,
    1: 1.3,
    2: 1.45,
    3: 1.6,
    4: 1.75,
    5: 1.9,
    6: 2.05,
    7: 2.2,
});
/**
 * @constant COMMON_CARDS - Maps Common cards from their in game English names to their config objects
 */
exports.COMMON_CARDS = Object.freeze({
    Damage: exports.DAMAGE_CARD,
    'Attack Speed': exports.ASPD_CARD,
    Health: exports.HEALTH_CARD,
    'Health Regen': exports.REGEN_CARD,
    Range: exports.RANGE_CARD,
    Cash: exports.CASH_CARD,
    Coins: exports.COIN_CARD,
    'Slow Aura': exports.SLOW_AURA_CARD,
    'Critical Chance': exports.CRIT_CHANCE_CARD,
    'Enemy Balance': exports.ENEMY_BALANCE_CARD,
    'Extra Defense': exports.EXTRA_DEFENSE_CARD,
    Fortress: exports.FORTRESS_CARD,
});
/**
 * @constant INTRO_SPRINT_CARD - Maps Intro Sprint Card stars to value, 0 is unequipped
 */
exports.INTRO_SPRINT_CARD = Object.freeze({
    0: 0,
    1: 20,
    2: 25,
    3: 30,
    4: 35,
    5: 40,
    6: 45,
    7: 50,
});
/**
 * @constant WAVE_SKIP_CARD - Maps Wave Skip Card stars to value, 0 is unequipped
 */
exports.WAVE_SKIP_CARD = Object.freeze({
    0: 0,
    1: 9,
    2: 10,
    3: 11,
    4: 13,
    5: 15,
    6: 17,
    7: 19,
});
/**
 * @constant CRIT_COIN_CARD - Maps Crit Coin Card stars to value, 0 is unequipped
 */
exports.CRIT_COIN_CARD = Object.freeze({
    0: 0,
    1: 15,
    2: 18,
    3: 21,
    4: 24,
    5: 27,
    6: 30,
    7: 33,
});
/**
 * @constant PLASMA_CANNON_CARD - Maps Plasma Cannon Card stars to value, 0 is unequipped
 */
exports.PLASMA_CANNON_CARD = Object.freeze({
    0: 0,
    1: 30,
    2: 34,
    3: 38,
    4: 42,
    5: 46,
    6: 50,
    7: 54,
});
/**
 * @constant EXTRA_ORB_CARD - Maps Extra Orb Card stars to value, 0 is unequipped
 */
exports.EXTRA_ORB_CARD = Object.freeze({
    0: 0,
    1: 0.3,
    2: 0.4,
    3: 0.5,
    4: 0.6,
    5: 0.7,
    6: 0.8,
    7: 0.9,
});
/**
 * @constant FREE_UPGRADES_CARD - Maps Free Upgrades Card stars to value, 0 is unequipped
 */
exports.FREE_UPGRADES_CARD = Object.freeze({
    0: 0,
    1: 4,
    2: 5,
    3: 6,
    4: 7,
    5: 8,
    6: 9,
    7: 10,
});
/**
 * @constant LAND_MINE_STUN_CARD - Maps Land Mine Stun Card stars to value, 0 is unequipped
 */
exports.LAND_MINE_STUN_CARD = Object.freeze({
    0: 0,
    1: 1.4,
    2: 1.8,
    3: 2.2,
    4: 2.6,
    5: 3,
    6: 3.4,
    7: 3.8,
});
/**
 * @constant RECOVERY_PACKAGE_CHANCE_CARD - Maps Recovery Package Chance Card stars to value, 0 is unequipped
 */
exports.RECOVERY_PACKAGE_CHANCE_CARD = Object.freeze({
    0: 0,
    1: 15,
    2: 18,
    3: 21,
    4: 24,
    5: 27,
    6: 30,
    7: 33,
});
/**
 * @constant RARE_CARDS - Maps Rare cards from their in game English names to their config objects
 */
exports.RARE_CARDS = Object.freeze({
    'Intro Sprint': exports.INTRO_SPRINT_CARD,
    'Wave Skip': exports.WAVE_SKIP_CARD,
    'Critical Coin': exports.CRIT_COIN_CARD,
    'Plasma Cannon': exports.PLASMA_CANNON_CARD,
    'Extra Orb': exports.EXTRA_ORB_CARD,
    'Free Upgrades': exports.FREE_UPGRADES_CARD,
    'Land Mine Stun': exports.LAND_MINE_STUN_CARD,
    'Recovery Package Chance': exports.RECOVERY_PACKAGE_CHANCE_CARD,
});
/**
 * @constant DEATH_RAY_CARD - Maps Death Ray Card stars to value, 0 is unequipped
 */
exports.DEATH_RAY_CARD = Object.freeze({
    0: 0,
    1: 2.3,
    2: 2.7,
    3: 3.1,
    4: 3.5,
    5: 3.9,
    6: 4.4,
    7: 4.9,
});
/**
 * @constant ENERGY_NET_CARD - Maps Energy Net Card stars to value, 0 is unequipped
 */
exports.ENERGY_NET_CARD = Object.freeze({
    0: 0,
    1: 2.5,
    2: 2.8,
    3: 3.1,
    4: 3.4,
    5: 3.7,
    6: 4.0,
    7: 4.3,
});
/**
 * @constant SUPER_TOWER_CARD - Maps Super Tower Card stars to value, 0 is unequipped
 */
exports.SUPER_TOWER_CARD = Object.freeze({
    0: 1,
    1: 2.5,
    2: 2.9,
    3: 3.3,
    4: 3.7,
    5: 4.1,
    6: 4.5,
    7: 5,
});
/**
 * @constant SECOND_WIND_CARD - Maps Second Wind Card stars to value, 0 is unequipped
 */
exports.SECOND_WIND_CARD = Object.freeze({
    0: 0,
    1: 10,
    2: 15,
    3: 20,
    4: 25,
    5: 30,
    6: 35,
    7: 40,
});
/**
 * @constant DEMON_MODE_CARD - Maps Demon Mode Card stars to value, 0 is unequipped
 */
exports.DEMON_MODE_CARD = Object.freeze({
    0: 0,
    1: 180,
    2: 200,
    3: 220,
    4: 240,
    5: 260,
    6: 280,
    7: 300,
});
/**
 * @constant ENERGY_SHIELD_CARD - Maps Energy Shield Card stars to value, 0 is unequipped
 */
exports.ENERGY_SHIELD_CARD = Object.freeze({
    0: 0,
    1: 20,
    2: 18,
    3: 16,
    4: 14,
    5: 12,
    6: 10,
    7: 8,
});
/**
 * @constant WAVE_ACCELERATOR_CARD - Maps Wave Accelerator Card stars to value, 0 is unequipped
 */
exports.WAVE_ACCELERATOR_CARD = Object.freeze({
    0: 0,
    1: 30,
    2: 34,
    3: 38,
    4: 42,
    5: 46,
    6: 50,
    7: 54,
});
/**
 * @constant BERSERKER_CARD - Maps Berserker Card stars to value, 0 is unequipped
 */
exports.BERSERKER_CARD = Object.freeze({
    0: 0,
    1: 1,
    2: 1.1,
    3: 1.3,
    4: 1.3,
    5: 1.4,
    6: 1.5,
    7: 1.6,
});
/**
 * @constant ULTIMATE_CRIT_CARD - Maps Ultimate Crit Card stars to value, 0 is unequipped
 */
exports.ULTIMATE_CRIT_CARD = Object.freeze({
    0: 0,
    1: 1,
    2: 1.33,
    3: 1.66,
    4: 2,
    5: 2.33,
    6: 2.66,
    7: 3,
});
/**
 * @constant EPIC_CARDS - Maps Epic cards from their in game English names to their config objects
 */
exports.EPIC_CARDS = Object.freeze({
    'Death Ray': exports.DEATH_RAY_CARD,
    'Energy Net': exports.ENERGY_NET_CARD,
    'Super Tower': exports.SUPER_TOWER_CARD,
    'Second Wind': exports.SECOND_WIND_CARD,
    'Demon Mode': exports.DEMON_MODE_CARD,
    'Energy Shield': exports.ENERGY_SHIELD_CARD,
    'Wave Accelerator': exports.WAVE_ACCELERATOR_CARD,
    Berserker: exports.BERSERKER_CARD,
    'Ultimate Crit': exports.ULTIMATE_CRIT_CARD,
});
/**
 * @constant CARDS - Maps All cards from their in game English names to their config objects
 */
exports.CARDS = Object.freeze({
    ...exports.COMMON_CARDS,
    ...exports.RARE_CARDS,
    ...exports.EPIC_CARDS,
});
/**
 * @constant LOCKED_CARDS - Maps Locked cards from their in game English names to the { tier, wave } milestone that they unlock
 */
exports.LOCKED_CARDS = Object.freeze({
    'Land Mine Stun': Object.freeze({
        tier: 7,
        wave: 250,
    }),
    'Recovery Package Chance': Object.freeze({
        tier: 2,
        wave: 750,
    }),
    'Ultimate Crit': Object.freeze({
        tier: 14,
        wave: 50,
    }),
});
