// tower-idle-toolkit/build/utils/formatTime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatTime = exports.YEAR = exports.DAY = exports.HOUR = exports.MINUTE = exports.SECOND = void 0;
exports.SECOND = 1;
exports.MINUTE = exports.SECOND * 60;
exports.HOUR = exports.MINUTE * 60;
exports.DAY = exports.HOUR * 24;
exports.YEAR = exports.DAY * 365;
/**
 * @function formatTime - Converts seconds to a nicely formatted string
 *
 * @param time - number in seconds
 * @returns string - nicely formatted as <> years, <> days, <>h, <>m, <>s, omitting values as necessary for clarity
 */
const formatTime = (time) => {
    const years = Math.floor(time / exports.YEAR);
    const afterYears = time % exports.YEAR;
    const days = Math.floor(afterYears / exports.DAY);
    const afterDays = afterYears % exports.DAY;
    const hours = Math.floor(afterDays / exports.HOUR);
    const afterHours = afterDays % exports.HOUR;
    const minutes = Math.floor(afterHours / exports.MINUTE);
    const seconds = Math.floor(afterHours % exports.MINUTE);
    const format = [];
    if (years > 0)
        format.push(`${years} years`);
    if (days > 0 || years > 0)
        format.push(`${days} days`);
    if (hours > 0)
        format.push(`${hours}h`);
    if (minutes > 0 || hours > 0)
        format.push(`${minutes}m`);
    format.push(`${seconds.toFixed(0)}s`);
    return format.join(', ');
};
exports.formatTime = formatTime;
